### PR TITLE
643: Adding domestic vrp validations

### PR DESCRIPTION
**Pull request by frlaura-jianu**
Issue: https://github.com/secureapigateway/secureapigateway/issues/643
Description: Added domestic vrp back-office validations.

### CODE REVIEW

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    protected List<OBError1> errors = new ArrayList<>();
+
+    public DomesticVrpValidationService clearErrors() {
+        this.errors.clear();
+        return this;
+    }
+
+    public List<OBError1> getErrors() {
+        return errors;
+    }
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java:52]: Error list never used, needs to be deleted if never used and the controller needs to be updated to call the validate directly without clear errors. `domesticVrpValidationService.clearErrors().validate(...)` to `domesticVrpValidationService.validate(...)`

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+    private OBRisk1Validator riskValidator;
+
+    public void validate(OBDomesticVRPInitiation initiation, OBDomesticVRPInstruction instruction, OBRisk1 risk, FRDomesticVrpRequest frDomesticVRPRequest, OBDomesticVRPControlParameters controlParameters, OBCashAccountCreditor3 requestCreditorAccount) throws OBErrorException {
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java:56]: The signature could be more simple passing only the consent and the request object
```java
public void validate(OBDomesticVRPConsentResponse consent, OBDomesticVRPRequest obDomesticVRPRequest)
...
```

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    //MaximumIndividualAmount - validation
+    private void validateMaximumIndividualAmount(FRDomesticVrpRequest consent, Double instructionAmount, String instructionCurrency, OBDomesticVRPControlParameters controlParameters) throws OBErrorException {
+        Double consentAmount = Double.valueOf(controlParameters.getMaximumIndividualAmount().getAmount());
+        String consentCurrency = controlParameters.getMaximumIndividualAmount().getCurrency();
+        if (!(instructionAmount.compareTo(consentAmount) == 0) || !(instructionCurrency.compareTo(consentCurrency) == 0)) {
+            throw new OBErrorException(
+                    OBRIErrorType.REQUEST_VRP_CONTROL_PARAMETERS_RULES);
+        }
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java:128]: the ASPSP must return an error with code UK.OBIE.Rules.FailsControlParameters and pass in the control parameter field that caused the error in the Field field of the error message.
Example:
```java
throw new OBErrorException(
                    OBRIErrorType.REQUEST_VRP_CONTROL_PARAMETERS_RULES,
                    VRPErrorControlParametersFields.RequestControlFields.MAX_INDIVIDUAL_AMOUNT,
                    VRPErrorControlParametersFields.ConsentControlFields.MAX_INDIVIDUAL_AMOUNT);
```

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    private void validateMaximumIndividualAmount(OBDomesticVRPInstruction instruction, FRDomesticVrpRequest consent, OBDomesticVRPControlParameters controlParameters) throws OBErrorException {
+        String instructionAmount = String.valueOf(instruction.getInstructedAmount().getAmount());
+        String instructionCurrency = String.valueOf(instruction.getInstructedAmount().getCurrency());
+        // TODO - next method must be implemented correctly depending on the control parameters rules set for the payment periodic limits and maximum
+        // individual ammount
+        validateMaximumIndividualAmount(consent, Double.valueOf(instructionAmount), instructionCurrency, controlParameters);
+    }
+
+    //MaximumIndividualAmount - validation
+    private void validateMaximumIndividualAmount(FRDomesticVrpRequest consent, Double instructionAmount, String instructionCurrency, OBDomesticVRPControlParameters controlParameters) throws OBErrorException {
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java:122]: `FRDomesticVrpRequest consent` never used.

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            throw new OBErrorException(OBRIErrorType.REQUEST_VRP_CREDITOR_ACCOUNT_NOT_SPECIFIED);
+        } else {
+            if (requestCreditorAccount.getIdentification() == null || requestCreditorAccount.getName() == null || requestCreditorAccount.getSchemeName() == null) {
+                throw new OBErrorException(OBRIErrorType.REQUEST_VRP_CREDITOR_ACCOUNT_NOT_SPECIFIED);
+            }
+        }
+    }
+
+    //controlParameters - validation
+    public void checkControlParameters(OBDomesticVRPInstruction instruction, FRDomesticVrpRequest consent, OBDomesticVRPControlParameters controlParameters) throws OBErrorException {
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java:109]: `FRDomesticVrpRequest consent` never used.

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        validateMaximumIndividualAmount(instruction, consent, controlParameters);
+    }
+
+    private void validateMaximumIndividualAmount(OBDomesticVRPInstruction instruction, FRDomesticVrpRequest consent, OBDomesticVRPControlParameters controlParameters) throws OBErrorException {
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java:113]: `FRDomesticVrpRequest consent` never used.

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/obie/payment/v3_1_8/vrp/DomesticVrpsApiController.java [hunk 7]
```diff
@@ -145,7 +156,44 @@ public ResponseEntity<OBDomesticVRPResponse> domesticVrpPost(
...
+                consentId
+        );
+
+        log.debug("Deserialized consent from IDM");
+
+        //get mandatory objects
+        OBDomesticVRPInitiation initiation = consent.getData().getInitiation();
+        OBDomesticVRPControlParameters controlParameters = consent.getData().getControlParameters();
+        OBRisk1 risk = consent.getRisk();
+        OBCashAccountCreditor3 requestCreditorAccount = consent.getData().getInitiation().getCreditorAccount();
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/obie/payment/v3_1_8/vrp/DomesticVrpsApiController.java:178]: Not necessary here, the validator signature could be more simple and delegate it to the validator.

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/obie/payment/v3_1_8/vrp/DomesticVrpsApiController.java [hunk 8]
```diff
@@ -145,7 +156,44 @@ public ResponseEntity<OBDomesticVRPResponse> domesticVrpPost(
...
+
+        // validate the consent against the instruction
+        try {
+            domesticVrpValidationService.clearErrors().validate(initiation, obDomesticVRPRequest.getData().getInstruction(), risk, frDomesticVRPRequest, controlParameters, requestCreditorAccount);
+        } catch (OBErrorException e) {
+            log.error("Got error during VRP validation: '{}'", e.getMessage());
+            throw new OBErrorResponseException(
+                    e.getObriErrorType().getHttpStatus(),
+                    OBRIErrorResponseCategory.REQUEST_FILTER,
+                    e.getOBError());
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/obie/payment/v3_1_8/vrp/DomesticVrpsApiController.java:192]: Return a generic error isn't right, the OBErrorResponseException returned needs to be builded with the OBErrorException returned by the validator with the proper OBErrorType returned by the validator.

**Review by jorgesanchezperez — CHANGES_REQUESTED** [review]
Please review the comments.

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java (hunk 3 above)
Comment (mariantiris) [reply securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/services/DomesticVrpValidationService.java:128]: Passed control parameter to error

**Review by jorgesanchezperez — APPROVED** [review]

**Pull request merged**